### PR TITLE
Add extra testing on headlistener for resubscribes; DefaultQueryCtx applies to all queries

### PR DESCRIPTION
**Pull request by samsondav**

### CODE REVIEW

**Review by PiotrTrzpil — APPROVED** [review]

File: core/services/headtracker/head_listener.go [hunk 1]
```diff
@@ -156,19 +153,21 @@ func (hl *HeadListener) receiveHeaders(ctx context.Context, handleNewHead func(c
 
 // subscribe periodically attempts to connect to the ethereum node via websocket.
 // It returns true on success, and false if cut short by a done request and did not connect.
+// NOT THREAD SAFE
 func (hl *HeadListener) subscribe() bool {
-	hl.sleeper.Reset()
+	hl.subscribeRetryBackoff.Reset()
+
 	for {
 		if err := hl.unsubscribeFromHead(); err != nil {
```
Comment (pinebit) [inline core/services/headtracker/head_listener.go:161]: `unsubscribeFromHead` always returns `null`, maybe we shall change the function to return nothing and remove this `if`?

File: core/services/headtracker/head_listener.go [hunk 2]
```diff
@@ -187,7 +186,9 @@ func (hl *HeadListener) subscribeToHead() error {
 
 	hl.headers = make(chan *eth.Head)
 
-	sub, err := hl.ethClient.SubscribeNewHead(context.Background(), hl.headers)
+	ctx, cancel := utils.ContextFromChan(hl.chStop)
```
Comment (pinebit) [inline core/services/headtracker/head_listener.go:189]: we have such the context created in `ListenForNewHeads`, can't we just pass that context all the way down to `subscribeToHead`?

**Review by pinebit — APPROVED** [review]

File: core/services/headtracker/head_listener.go (hunk 1 above)
Comment (samsondav) [reply core/services/headtracker/head_listener.go:161]: Good catch! This code is very old and a bit wonky in places, will fix.

**Review by pinebit — APPROVED** [review]
LGTM

**Review by pinebit — DISMISSED** [review]

File: core/services/eth/chain_id_sub.go [hunk 3]
```diff
@@ -46,43 +48,43 @@ func (c *chainIDSubForwarder) start(sub ethereum.Subscription, err error) error
...
-
-	close(c.err)
+	// tell forwardLoop to unsubscribe
+	select {
+	case c.unSub <- struct{}{}:
+	default:
+		// already triggered
+	}
+	// wait for forwardLoop to complete
+	<-c.done
```
Comment (jmank88) [inline core/services/eth/chain_id_sub.go:87]: Now we never block sending on c.unSub, even if called multiple times.

**Review by jmank88 — APPROVED** [review]

**Pull request merged**